ara/com/ipc/src/ux_socket.cpp

```diff
@@ -0,0 +1,89 @@
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <unistd.h>
+
+#include <iostream>
+
+#include "ux_socket.h"
+#include "socket.h"
+
+namespace ara {
+namespace com {
+namespace ipc {
+
+inline void exit_error(char const* msg)
+{
+    if (msg != nullptr)
+        std::cerr << msg << '\n';
+    exit(EXIT_FAILURE);
+}
+
+UxSocket::UxSocket(ISocket* isock_impl) 
+    : sock_impl_(isock_impl)
+{
+    sock_fd_ = 
+            sock_impl_->socket(AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC | SOCK_NONBLOCK, 0);
+
+    if (sock_fd_ == -1) {
+        throw std::runtime_error("UxSocket open failure(-1)\n");
+    }
+}
+
+UxSocket::~UxSocket()
+{
+    if (sock_impl_->close(sock_fd_) == -1)
+        exit_error("UxSocket close() failure(-1)\n");
+}
+
+void UxSocket::configure_sockaddr_un(sockaddr_un& addr, char const* path) const noexcept
+{
+    memset(&addr, 0, sizeof(sockaddr_un));
+    addr.sun_family = AF_UNIX;
+    strncpy(addr.sun_path, path, sizeof(sockaddr_un) - 1);
```

Comment:
Why is it `sizeof(sockaddr_un) - 1` instead of `strlen(path)`?

Reply:
`char *strncpy( char *dest, const char *src, size_t count );'
`strncpy` copies at most count characters to the `dest` array stopping at first null character. Since memset sets all values of `struct addr` to 0 this action ensures that `sun_path` always has terminating null byte, even in case of `path` argument being larger that `sun_path` array.
For better explanation see: M. Kerrisk - The Linux Programming Interface (2010, p. 1166).

Reply:
Ok, I understand now- good to know. I was just wondering why do we copy content of one variable with size of another one.